File: frontend/src/services/blogService.js
import api from './api';

const BLOG_ENDPOINTS = {
  // Public endpoints
  LIST: '/api/blogs',
  LIST_ALL: '/api/blogs/all',
  DETAIL: (id) => `/api/blogs/${id}`,
  SLUG: (slug) => `/api/blogs/slug/${slug}`,
  CATEGORY: (category) => `/api/blogs/category/${category}`,
  SEARCH: '/api/blogs/search',
  RECENT: '/api/blogs/recent',
  CATEGORIES: '/api/blogs/categories',
  IMAGE: (id) => `/api/blogs/${id}/image`,
  SUBSCRIBE: '/api/blogs/subscribe',
  UNSUBSCRIBE: (token) => `/api/blogs/unsubscribe/${token}`,
  SUBSCRIBER_COUNT: '/api/blogs/subscribers/count',

  // Admin endpoints
  ADMIN_LIST: '/api/blogs/admin/all',
  ADMIN_DETAIL: (id) => `/api/blogs/admin/${id}`,
  ADMIN_CREATE: '/api/blogs/admin',
  ADMIN_UPDATE: (id) => `/api/blogs/admin/${id}`,
  ADMIN_DELETE: (id) => `/api/blogs/admin/${id}`,
  ADMIN_STATUS: (id) => `/api/blogs/admin/${id}/status`,
  ADMIN_IMAGE: (id) => `/api/blogs/admin/${id}/image`,
  ADMIN_SUBSCRIBERS: '/api/blogs/admin/subscribers',
};

// ==================== PUBLIC METHODS ====================

/**
 * Get published blogs with pagination
 * @param {number} page - Page number (0-indexed)
 * @param {number} size - Page size
 * @returns {Promise<Object>} Paginated blog list
 */
export async function getPublishedBlogs(page = 0, size = 10) {
  const response = await api.get(BLOG_ENDPOINTS.LIST, { params: { page, size } });
  return response.data;
}

/**
 * Get all published blogs (no pagination)
 * @returns {Promise<Array>} List of published blogs
 */
export async function getAllBlogs() {
  const response = await api.get(BLOG_ENDPOINTS.LIST_ALL);
  return response.data;
}

/**
 * Get blog by ID (public - published only)
 * @param {number|string} id - Blog ID
 * @returns {Promise<Object>} Blog post details
 */
export async function getBlog(id) {
  const response = await api.get(BLOG_ENDPOINTS.DETAIL(id));
  return response.data;
}

/**
 * Get blog by URL slug
 * @param {string} slug - URL slug
 * @returns {Promise<Object>} Blog post details
 */
export async function getBlogBySlug(slug) {
  const response = await api.get(BLOG_ENDPOINTS.SLUG(slug));
  return response.data;
}

/**
 * Get blogs by category
 * @param {string} category - Category name
 * @returns {Promise<Array>} List of blogs in category
 */
export async function getBlogsByCategory(category) {
  const response = await api.get(BLOG_ENDPOINTS.CATEGORY(category));
  return response.data;
}

/**
 * Search blogs
 * @param {string} query - Search query
 * @returns {Promise<Array>} Search results
 */
export async function searchBlogs(query) {
  const response = await api.get(BLOG_ENDPOINTS.SEARCH, { params: { q: query } });
  return response.data;
}

/**
 * Get recent blogs
 * @returns {Promise<Array>} List of recent blogs
 */
export async function getRecentBlogs() {
  const response = await api.get(BLOG_ENDPOINTS.RECENT);
  return response.data;
}

/**
 * Get categories with count
 * @returns {Promise<Array>} List of categories with counts
 */
export async function getCategoriesWithCount() {
  const response = await api.get(BLOG_ENDPOINTS.CATEGORIES);
  return response.data;
}

/**
 * Get blog featured image URL
 * @param {number|string} id - Blog ID
 * @returns {string} Image URL
 */
export function getBlogImageUrl(id) {
  return `${api.defaults.baseURL || ''}${BLOG_ENDPOINTS.IMAGE(id)}`;
}

// ==================== SUBSCRIPTION METHODS ====================

/**
 * Subscribe to blog newsletter
 * @param {string} email - Email address
 * @param {string} name - Subscriber name
 * @returns {Promise<Object>} Subscription result
 */
export async function subscribe(email, name) {
  const response = await api.post(BLOG_ENDPOINTS.SUBSCRIBE, { email, name });
  return response.data;
}

/**
 * Unsubscribe from blog newsletter
 * @param {string} token - Unsubscribe token
 * @returns {Promise<Object>} Unsubscribe result
 */
export async function unsubscribe(token) {
  const response = await api.get(BLOG_ENDPOINTS.UNSUBSCRIBE(token));
  return response.data;
}

/**
 * Get subscriber count
 * @returns {Promise<Object>} Subscriber count
 */
export async function getSubscriberCount() {
  const response = await api.get(BLOG_ENDPOINTS.SUBSCRIBER_COUNT);
  return response.data;
}

// ==================== ADMIN METHODS ====================

/**
 * Get all blogs (admin)
 * @returns {Promise<Array>} List of all blogs
 */
export async function getAdminBlogs() {
  const response = await api.get(BLOG_ENDPOINTS.ADMIN_LIST);
  return response.data;
}

/**
 * Get blog by ID (admin - any status)
 * @param {number|string} id - Blog ID
 * @returns {Promise<Object>} Blog post details
 */
export async function getAdminBlog(id) {
  const response = await api.get(BLOG_ENDPOINTS.ADMIN_DETAIL(id));
  return response.data;
}

/**
 * Create a new blog post (Admin/Author only)
 * @param {Object} blogData - Blog post data
 * @returns {Promise<Object>} Created blog post
 */
export async function createBlog(blogData) {
  const response = await api.post(BLOG_ENDPOINTS.ADMIN_CREATE, blogData);
  return response.data;
}

/**
 * Update a blog post (Admin/Author only)
 * @param {number|string} id - Blog ID
 * @param {Object} blogData - Updated blog data
 * @returns {Promise<Object>} Updated blog post
 */
export async function updateBlog(id, blogData) {
  const response = await api.put(BLOG_ENDPOINTS.ADMIN_UPDATE(id), blogData);
  return response.data;
}

/**
 * Delete a blog post (Admin only)
 * @param {number|string} id - Blog ID
 * @returns {Promise<void>}
 */
export async function deleteBlog(id) {
  await api.delete(BLOG_ENDPOINTS.ADMIN_DELETE(id));
}

/**
 * Update blog status
 * @param {number|string} id - Blog ID
 * @param {string} status - New status (DRAFT, PUBLISHED, SCHEDULED, ARCHIVED)
 * @returns {Promise<Object>} Updated blog
 */
export async function updateBlogStatus(id, status) {
  const response = await api.patch(BLOG_ENDPOINTS.ADMIN_STATUS(id), { status });
  return response.data;
}

/**
 * Upload featured image
 * @param {number|string} id - Blog ID
 * @param {File} file - Image file
 * @returns {Promise<Object>} Updated blog
 */
export async function uploadFeaturedImage(id, file) {
  const formData = new FormData();
  formData.append('file', file);
  const response = await api.post(BLOG_ENDPOINTS.ADMIN_IMAGE(id), formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data;
}

/**
 * Get admin blog image URL
 * @param {number|string} id - Blog ID
 * @returns {string} Image URL
 */
export function getAdminBlogImageUrl(id) {
  return `${api.defaults.baseURL || ''}${BLOG_ENDPOINTS.ADMIN_IMAGE(id)}`;
}

/**
 * Get all active subscribers (admin)
 * @returns {Promise<Array>} List of active subscribers
 */
export async function getActiveSubscribers() {
  const response = await api.get(BLOG_ENDPOINTS.ADMIN_SUBSCRIBERS);
  return response.data;
}

export default {
  // Public
  getPublishedBlogs,
  getAllBlogs,
  getBlog,
  getBlogBySlug,
  getBlogsByCategory,
  searchBlogs,
  getRecentBlogs,
  getCategoriesWithCount,
  getBlogImageUrl,
  // Subscription
  subscribe,
  unsubscribe,
  getSubscriberCount,
  // Admin
  getAdminBlogs,
  getAdminBlog,
  createBlog,
  updateBlog,
  deleteBlog,
  updateBlogStatus,
  uploadFeaturedImage,
  getAdminBlogImageUrl,
  getActiveSubscribers,
};

File: src/main/java/com/uyirgene/blog/BlogService.java
package com.uyirgene.blog;

import com.uyirgene.course.MailService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
public class BlogService {

    private final BlogRepository blogRepo;
    private final BlogSubscriptionRepository subscriptionRepo;
    private final MailService mailService;

    // ==================== PUBLIC BLOG METHODS ====================

    /**
     * Get all published public blogs
     */
    public List<Blog> getPublishedBlogs() {
        return blogRepo.findByStatusAndVisibilityOrderByPublishDateDesc(
                Blog.BlogStatus.PUBLISHED, Blog.BlogVisibility.PUBLIC);
    }

    /**
     * Get published blogs with pagination
     */
    public Page<Blog> getPublishedBlogs(Pageable pageable) {
        return blogRepo.findByStatusAndVisibilityOrderByPublishDateDesc(
                Blog.BlogStatus.PUBLISHED, Blog.BlogVisibility.PUBLIC, pageable);
    }

    /**
     * Get blog by URL slug (public)
     */
    public Optional<Blog> getBlogBySlug(String slug) {
        Optional<Blog> blogOpt = blogRepo.findByUrlSlug(slug);
        // Increment view count
        blogOpt.ifPresent(blog -> {
            blog.setViewCount(blog.getViewCount() + 1);
            blogRepo.save(blog);
        });
        return blogOpt;
    }

    /**
     * Get blog by ID (public - only if published and public)
     */
    public Optional<Blog> getPublicBlogById(Long id) {
        return blogRepo.findById(id)
                .filter(b -> b.getStatus() == Blog.BlogStatus.PUBLISHED
                        && b.getVisibility() == Blog.BlogVisibility.PUBLIC);
    }

    /**
     * Get blogs by category
     */
    public List<Blog> getBlogsByCategory(String category) {
        return blogRepo.findByCategoryAndStatusAndVisibilityOrderByPublishDateDesc(
                category, Blog.BlogStatus.PUBLISHED, Blog.BlogVisibility.PUBLIC);
    }

    /**
     * Search blogs
     */
    public List<Blog> searchBlogs(String query) {
        return blogRepo.searchBlogs(query, Blog.BlogStatus.PUBLISHED, Blog.BlogVisibility.PUBLIC);
    }

    /**
     * Get recent blogs
     */
    public List<Blog> getRecentBlogs() {
        return blogRepo.findTop5ByStatusAndVisibilityOrderByPublishDateDesc(
                Blog.BlogStatus.PUBLISHED, Blog.BlogVisibility.PUBLIC);
    }

    /**
     * Get categories with count
     */
    public List<Object[]> getCategoriesWithCount() {
        return blogRepo.getCategoriesWithCount();
    }

    // ==================== ADMIN BLOG METHODS ====================

    /**
     * Get all blogs (admin)
     */
    public List<Blog> getAllBlogs() {
        return blogRepo.findAll();
    }

    /**
     * Get blog by ID (admin - any status)
     */
    public Optional<Blog> getBlogById(Long id) {
        return blogRepo.findById(id);
    }

    /**
     * Create a new blog
     */
    @Transactional
    public Blog createBlog(Blog blog) {
        // Generate unique slug
        String baseSlug = Blog.generateSlug(blog.getTitle());
        String slug = baseSlug;
        int counter = 1;
        while (blogRepo.existsByUrlSlug(slug)) {
            slug = baseSlug + "-" + counter++;
        }
        blog.setUrlSlug(slug);

        // Set publish date if publishing
        if (blog.getStatus() == Blog.BlogStatus.PUBLISHED && blog.getPublishDate() == null) {
            blog.setPublishDate(LocalDateTime.now());
        }

        Blog saved = blogRepo.save(blog);
        log.info("Blog created: {} (slug: {})", saved.getTitle(), saved.getUrlSlug());

        // Notify subscribers if published
        if (saved.getStatus() == Blog.BlogStatus.PUBLISHED) {
            notifySubscribers(saved);
        }

        return saved;
    }

    /**
     * Update an existing blog
     */
    @Transactional
    public Blog updateBlog(Long id, Blog blogData) {
        Blog existing = blogRepo.findById(id)
                .orElseThrow(() -> new RuntimeException("Blog not found: " + id));

        boolean wasNotPublished = existing.getStatus() != Blog.BlogStatus.PUBLISHED;

        // Update fields (preserve createdAt and urlSlug unless explicitly changed)
        existing.setTitle(blogData.getTitle());
        existing.setShortDescription(blogData.getShortDescription());
        existing.setContent(blogData.getContent());
        existing.setAuthorName(blogData.getAuthorName());
        existing.setCategory(blogData.getCategory());
        existing.setTags(blogData.getTags());
        existing.setStatus(blogData.getStatus());
        existing.setVisibility(blogData.getVisibility());
        existing.setLanguage(blogData.getLanguage());

        // Update slug only if explicitly provided and different
        if (blogData.getUrlSlug() != null && !blogData.getUrlSlug().equals(existing.getUrlSlug())) {
            if (!blogRepo.existsByUrlSlug(blogData.getUrlSlug())) {
                existing.setUrlSlug(blogData.getUrlSlug());
            }
        }

        // Set publish date if transitioning to published
        if (blogData.getStatus() == Blog.BlogStatus.PUBLISHED) {
            if (existing.getPublishDate() == null) {
                existing.setPublishDate(LocalDateTime.now());
            }
        }

        // Handle scheduled publish date
        if (blogData.getStatus() == Blog.BlogStatus.SCHEDULED && blogData.getPublishDate() != null) {
            existing.setPublishDate(blogData.getPublishDate());
        }

        Blog saved = blogRepo.save(existing);
        log.info("Blog updated: {} (id: {})", saved.getTitle(), saved.getId());

        // Notify subscribers if just published
        if (wasNotPublished && saved.getStatus() == Blog.BlogStatus.PUBLISHED) {
            notifySubscribers(saved);
        }

        return saved;
    }

    /**
     * Delete a blog
     */
    @Transactional
    public void deleteBlog(Long id) {
        Blog blog = blogRepo.findById(id)
                .orElseThrow(() -> new RuntimeException("Blog not found: " + id));
        blogRepo.delete(blog);
        log.info("Blog deleted: {} (id: {})", blog.getTitle(), id);
    }

    /**
     * Update blog status
     */
    @Transactional
    public Blog updateStatus(Long id, Blog.BlogStatus status) {
        Blog blog = blogRepo.findById(id)
                .orElseThrow(() -> new RuntimeException("Blog not found: " + id));

        boolean wasNotPublished = blog.getStatus() != Blog.BlogStatus.PUBLISHED;
        blog.setStatus(status);

        if (status == Blog.BlogStatus.PUBLISHED && blog.getPublishDate() == null) {
            blog.setPublishDate(LocalDateTime.now());
        }

        Blog saved = blogRepo.save(blog);

        // Notify subscribers if just published
        if (wasNotPublished && status == Blog.BlogStatus.PUBLISHED) {
            notifySubscribers(saved);
        }

        return saved;
    }

    /**
     * Update featured image
     */
    @Transactional
    public Blog updateFeaturedImage(Long id, byte[] image, String contentType) {
        Blog blog = blogRepo.findById(id)
                .orElseThrow(() -> new RuntimeException("Blog not found: " + id));
        blog.setFeaturedImage(image);
        blog.setFeaturedImageContentType(contentType);
        return blogRepo.save(blog);
    }

    // ==================== SUBSCRIPTION METHODS ====================

    /**
     * Subscribe to blog notifications
     */
    @Transactional
    public BlogSubscription subscribe(String email, String name) {
        Optional<BlogSubscription> existing = subscriptionRepo.findByEmail(email);
        if (existing.isPresent()) {
            BlogSubscription sub = existing.get();
            if (!sub.getActive()) {
                sub.setActive(true);
                sub.setUnsubscribedAt(null);
                sub.setName(name);
                return subscriptionRepo.save(sub);
            }
            return sub; // Already subscribed
        }

        BlogSubscription subscription = BlogSubscription.builder()
                .email(email)
                .name(name)
                .active(true)
                .build();
        return subscriptionRepo.save(subscription);
    }

    /**
     * Unsubscribe using token
     */
    @Transactional
    public boolean unsubscribe(String token) {
        Optional<BlogSubscription> subOpt = subscriptionRepo.findByUnsubscribeToken(token);
        if (subOpt.isPresent()) {
            BlogSubscription sub = subOpt.get();
            sub.setActive(false);
            sub.setUnsubscribedAt(LocalDateTime.now());
            subscriptionRepo.save(sub);
            return true;
        }
        return false;
    }

    /**
     * Get active subscribers count
     */
    public long getSubscriberCount() {
        return subscriptionRepo.countByActiveTrue();
    }

    /**
     * Get all active subscribers (admin)
     */
    public List<BlogSubscription> getActiveSubscribers() {
        return subscriptionRepo.findByActiveTrue();
    }

    /**
     * Notify subscribers about new blog post
     */
    private void notifySubscribers(Blog blog) {
        List<BlogSubscription> subscribers = subscriptionRepo.findByActiveTrue();
        if (subscribers.isEmpty()) {
            log.info("No active subscribers to notify for blog: {}", blog.getTitle());
            return;
        }

        log.info("Notifying {} subscribers about new blog: {}", subscribers.size(), blog.getTitle());

        // Send email notifications asynchronously
        mailService.sendNewBlogNotification(subscribers, blog);
    }
}
